apps/admin/backend/src/cast_vote_records.ts

```diff
@@ -0,0 +1,283 @@
+      );
+      const addCastVoteRecordResult = store.addCastVoteRecordFileEntry({
+        ballotId: castVoteRecord.UniqueId as BallotId,
+        cvr: {
+          ballotStyleId: castVoteRecord.BallotStyleId,
+          batchId: castVoteRecord.BatchId,
+          card: castVoteRecordBallotSheetId
+            ? { type: 'hmpb', sheetNumber: castVoteRecordBallotSheetId }
+            : { type: 'bmd' },
+          precinctId: castVoteRecord.BallotStyleUnitId,
+          votes,
+          votingMethod: castVoteRecord.vxBallotType,
+        },
+        cvrFileId: importId,
+        electionId,
+      });
+      if (addCastVoteRecordResult.isErr()) {
+        return err({
+          ...addCastVoteRecordResult.err(),
+          index: castVoteRecordIndex,
+        });
+      }
+      const { cvrId: castVoteRecordId, isNew: isCastVoteRecordNew } =
+        addCastVoteRecordResult.ok();
+
+      if (isCastVoteRecordNew) {
+        const hmpbCastVoteRecordWriteIns = castVoteRecordWriteIns.filter(
+          (castVoteRecordWriteIn) => castVoteRecordWriteIn.side
+        );
+        if (hmpbCastVoteRecordWriteIns.length > 0) {
+          // Guaranteed to exist given validation in readCastVoteRecordExport
+          assert(referencedFiles !== undefined);
+
+          for (const i of [0, 1] as const) {
+            const imageData = await fs.readFile(
+              referencedFiles.imageFilePaths[i]
+            );
+            const parseLayoutResult = safeParseJson(
+              await fs.readFile(referencedFiles.layoutFilePaths[i], 'utf8'),
+              BallotPageLayoutSchema
+            );
+            if (parseLayoutResult.isErr()) {
+              return err({
+                type: 'invalid-cast-vote-record',
+                subType: 'layout-parse-error',
+                index: castVoteRecordIndex,
+              });
+            }
+            store.addBallotImage({
+              cvrId: castVoteRecordId,
+              imageData,
+              pageLayout: parseLayoutResult.ok(),
+              side: (['front', 'back'] as const)[i],
+            });
+          }
+
+          for (const hmpbCastVoteRecordWriteIn of hmpbCastVoteRecordWriteIns) {
+            store.addWriteIn({
+              castVoteRecordId,
+              contestId: hmpbCastVoteRecordWriteIn.contestId,
+              electionId,
+              optionId: hmpbCastVoteRecordWriteIn.optionId,
+              side: assertDefined(hmpbCastVoteRecordWriteIn.side),
+            });
+          }
+        }
+      }
+
+      if (isCastVoteRecordNew) {
+        newlyAdded += 1;
+      } else {
+        alreadyPresent += 1;
+      }
+      precinctIds.add(castVoteRecord.BallotStyleUnitId);
+
+      castVoteRecordIndex += 1;
+    }
+
+    // TODO: Calculate the precinct list before iterating through cast vote records, once there is
+    // only one geopolitical unit per batch
+    store.updateCastVoteRecordFileRecord({
+      id: importId,
+      precinctIds,
+    });
```

Comment:
Now that we have a metadata file, I'm hoping we can eventually get the precinct list in the file up front to avoid this.

Reply:
👍 Opened an issue for myself so that I don't forget to revisit this